### ISSUE TITLE
Two Issues: 1 with Coarse Transformer, 1 with Validation Loss

### ISSUE THREAD

**Issue by kauschie** [lucidrains/audiolm-pytorch#287]
Hey all, 

I've been working with this library now since the end of December trying to get things up and running and trained on a really simple musical audio data set that is even more simplified than midi songs.. they're just some random beeps and boops with each training sample being 2s in length. However, most of my initial failures have actually been a result of of a couple of bugs that I've stumbled upon. I should state upfront that I am using Encodec as the audio codec in case it ends up making a difference. 

1) When training, I noticed that there were huge discrepancies with my step loss. I use a signal handler in my script to capture SIGINT so that i could pause training by first saving the checkpoint before exiting. However, when loading the saved checkpoint from my terminated session, the initial step loss calculated was drastically different from when I terminated the session. At first, I believed it to be due to how I was saving and loading the model, but I was able to rule that out with more investigation and testing. After closer observation, the next calculated step loss can be >1.0 different after a validation loss was calculated and this was while I'd been plateauing around a certain value for hundreds of steps. I thought that was quite strange so I tested it further and my suspicions were correct. If the validation loss isn't calculated then the output step loss stays consistent to it's current trend while training. Additionally, there were no more discrepancies with the loss value with saving and loading checkpoints if validation losses aren't calculated. Though, now I don't have an idea of my model's performance and fit but as long as I can hear what's going on, I can at least judge it's progress subjectively. I'm not sure if this behavior is desired or not, but I believe this is occurring due to line 1274 in train.py:

`with torch.no_grad():`

which allows for gradient modifications in place while the other trainers are using `torch.inference_mode()` which does not. I'm still very much a novice though so, again, perhaps its desired and I'm mistakenly complaining about my model converging quicker than I had anticipated. However, upon restarting training, the loss was right where the last step loss was prior to the validation loss calculation. So it seems like this is undesirable. It also interestingly begs the question:

IF the step loss changes after the validation loss is calculated, and IF every step afterwards is very close numerically in value to this newly offset step loss,  and IF each subsequently calculated validation loss is calculated and the new loss is similarly offset accordingly,

then the step loss and validation loss output logged to the console are just offsets of the actual? 

I'm not quite sure of what that could mean other than possibly the loss is supposed to be influenced by the validation set and what I'm observing could be attributed to the learning rate being different from save to load but the following save and loads from their respective classes:

```
def save(self, path):
    pkg = dict(
    model = self.accelerator.get_state_dict(self.transformer),
    optim = self.optim.state_dict(),
    version = __version__
)
```
 
```torch.save(pkg, path)```

```
def load(self):
    ...
    self.optim.load_state_dict(pkg['optim'])
```

makes it seem like things are being loaded correctly on that front so I'm not quire sure what else to make of it if that isn't the cause.

2) The second issue that I found pertains to training the coarse transformer and this one was the bane of my existence for a while. I was attempting to over-fit my models on one song as I had seen suggested in the discussions. Though, I was repeatedly getting garbage audio with no recourse. Nothing I was trying was working and I had almost given up. After all of my models were trained, I used the following few line to generate audio during inference similar to what's shown in the example:

```
generated_wave = audiolm(batch_size=1)
sample_rate = 24000  # Example: 24kHz
output_file = "generated_audio.wav"
torchaudio.save(output_file, generated_wave.cpu(), sample_rate)
```

and i would use torch audio to save file. However, tqdm was indicating some really strange values as far as how many tokens it's attempting to generate no matter what max length I tossed in to attempt to change this. To be honest, it was as if the max length argument had no effect on anything. However, the quick fix presented itself though I still haven't traced the reason for it yet. 

I was still trying to get the model to produce audio at this point so I had been over-fitting it on one song; training of the coarse transformer finished relatively quickly which didn't require that I pause training so I just let it run until completion. Every time though, I could not get anything besides nails on a chalkboard to come out. One day when I was testing things, I terminated the coarse transformer prematurely and then restarted training from the saved checkpoint. When doing inference with this coarse transformer the same return value from the above lines of code was no longer a tensor and was instead a list of tensors 1 big thus crashing my script which lacked any kind of exception handling. Somehow the return value from an identical function call is different?! Not only that, there were now indications that semantic token generation was actually giving a darn about the max_length that i was intermittently trying to pass in. 

I eventually landed on the belief that I must have some kind of hyperparameter mismatch, chalked it up to the belief that I logged the wrong value somehow in my own logging functions, scrapped all of the models and just started over again. However, I reanalyzed the training and inference scripts and everything was a match so I retrained the same models only this time they didn't throw any exceptions, it gave me the tensor as the output, the semantic token numers reported by tqdm returned to being garbage, and resultant music hurt my ears. Eventually, I started getting lists again and I finally listened to the tensor contained in the list and heard the audio track it was trained on for the first time. 

TLDR - Ultimately, it's required to pause training and then restart it in order for inference to work. It must also be the case that loading the model's saved checkpoint during training results in correctly initializing the coarse transformer which later gets saved correctly. Also, I noticed that its the coarse transformer specifically that needed to be terminated and then having training re-initiated, followed by the model being saved again in order to get the output of inference to be a list with valid audio. At this point, I terminate and re-instantiate training for all three of the transformer right in the beginning for superstitious reasons even though I've narrowed it down to the coarse transformer. I haven't been able to trace it to the code responsible, or test the effect of when when this save-load-save process occurs in the training cycle. I found the fix and I moved on. However, I feel like this bit of info can surely help someone else who's pulling their hair out if it never gets resolved.

Additionally, it's my understanding that since I've trained on 2 seconds of audio, it's likely going to be spitting out ~2 seconds of generated audio which I think has to do with including the EOS token in training. Currently, I'm slicing off my own seed from the tail end of the tensor output and sending it back in for inference multiple times as an input prime wave. Though, by the end of the elongated sequence, this hacky method doesn't generate anything intelligible relating to the beginning of the initial audio sequence. Additionally, the unique portion of the output is now some fraction of the ~2 seconds due to needing an overlap with the seed. Inference also takes quite a bit longer using this method since inference needs to be conducted 6-7 times from start to finish in order to generate one 10s clip. Is there a better method than this like perhaps taking the slice directly from the semantic tokens and re-routing it through the semantic transformer multiple times? I also thought of re-processing my training data to consist of longer clips from the same exact music and retraining as a form of curriculum based learning. Was looking for the best solution available in this situation as I would really prefer the clips be 2 seconds in length due to hardware limitations / training time.

Excellent resource though and I appreciate all your hard work and any insight you could provide
